crates/server/src/datastore/postgres/mod.rs

```diff
@@ -0,0 +1,774 @@
+        .await?
+        .into_iter()
+        .map(|(record_id, c)| {
+            ProtoEnvelope::from_protobuf(c).map_err(|e| DataStoreError::InvalidRecordContents {
+                record_id: record_id.0.into(),
+                message: e.to_string(),
+            })
+        })
+        .collect::<Result<_, _>>()
+}
+
+async fn insert_record<V>(
+    conn: &mut AsyncPgConnection,
+    log_id: &LogId,
+    name: Option<&str>,
+    record_id: &RecordId,
+    record: &ProtoEnvelope<V::Record>,
+    sources: &[ContentSource],
+) -> Result<(), DataStoreError>
+where
+    V: Validator + 'static,
+    <V as Validator>::Error: ToString + Send + Sync,
+    DataStoreError: From<<V as Validator>::Error>,
+{
+    conn.transaction::<_, DataStoreError, _>(|conn| {
+        async move {
+            // Unfortunately, this cannot be done with an ON CONFLICT DO NOTHING clause as
+            // data cannot be returned; so just do a query for the log id and insert if it doesn't exist.
+            let log_id = match schema::logs::table
+                .select(schema::logs::id)
+                .filter(schema::logs::log_id.eq(TextRef(log_id)))
+                .first::<i32>(conn)
+                .await
+                .optional()?
+            {
+                Some(id) => id,
+                None => diesel::insert_into(schema::logs::table)
+                    .values(NewLog {
+                        log_id: TextRef(log_id),
+                        name,
+                        validator: &Json(V::default()),
+                    })
+                    .returning(schema::logs::id)
+                    .get_result::<i32>(conn)
+                    .await
+                    .map_err(|e| match e {
+                        diesel::result::Error::DatabaseError(
+                            DatabaseErrorKind::UniqueViolation,
+                            _,
+                        ) => DataStoreError::Conflict,
+                        e => e.into(),
+                    })?,
+            };
+
+            let id = diesel::insert_into(schema::records::table)
+                .values(NewRecord {
+                    log_id,
+                    record_id: TextRef(record_id),
+                    content: &record.to_protobuf(),
+                })
+                .returning(schema::records::id)
+                .get_result::<i32>(conn)
+                .await
+                .map_err(|e| match e {
+                    diesel::result::Error::DatabaseError(DatabaseErrorKind::UniqueViolation, _) => {
+                        DataStoreError::Conflict
+                    }
+                    e => e.into(),
+                })?;
+
+            for source in sources {
+                let (kind, url) = match &source.kind {
+                    ContentSourceKind::HttpAnonymous { url } => {
+                        (SourceKind::Http, Some(url.as_str()))
+                    }
+                };
+
+                diesel::insert_into(schema::sources::table)
```

Comment:
Is there not an insert-many?

Reply:
Apparently `values` can take an iterator for multiple value inserts, so I'll fix this.